nuxeo/nuxeo.py

```diff
@@ -443,54 +469,64 @@ def request_authentication_token(
         url = self.base_url + 'authentication/token?'
         url += urlencode(parameters)
 
-        headers = self._get_common_headers()
-        cookies = self._get_cookies()
-        self.trace('Calling %s with headers %r and cookies %r',
-                   url, headers, cookies)
-        req = urllib2.Request(url, headers=headers)
-        try:
-            token = self.opener.open(req, timeout=self.timeout).read()
-        except urllib2.HTTPError as e:
-            if e.code == 401 or e.code == 403:
-                raise Unauthorized(self.base_url, self.user_id, e.code)
-            elif e.code == 404:
-                # Token based auth is not supported by this server
-                return None
-            else:
-                e.msg = err + ': HTTP error %d' % e.code
-                raise e
-        except Exception as e:
-            if hasattr(e, 'msg'):
-                e.msg = err + ': ' + e.msg
-            raise
-        cookies = self._get_cookies()
-        self.trace('Got token %r with cookies %r', token, cookies)
+        token = self.send(url).text
+
         # Use the (potentially re-newed) token from now on
         if not revoke:
             self._update_auth(token=token)
         return token
 
+    def send(self, url, method='GET', data=None, extra_headers=None, timeout=None):
+        if method not in ('GET', 'HEAD', 'POST', 'PUT',
+                          'DELETE', 'CONNECT', 'OPTIONS', 'TRACE'):
+            raise ValueError('method parameter is not a valid HTTP method.')
+
+        timeout = self.timeout if not timeout or timeout == -1 else timeout
+        headers = self.headers(extra_headers)
+
+        logger.debug('Calling {!s} with headers {!r} and cookies {!r}'.format(
+            url, headers, self._get_cookies()))
+
+        try:
+            resp = self._session.request(url=url, method=method, headers=headers,
+                                         data=data, timeout=timeout)
+            resp.raise_for_status()
+        except HTTPError as e:
+            self._log_details(e)
+            code = e.response.status_code
+            if code in (401, 403):
+                raise Unauthorized(self.base_url, self.user_id, code)
+            else:
+                raise e
+        except Exception as e:
+            self._log_details(e)
+            raise e
+
+        if ('content-length' in resp.headers
+            and int(resp.headers['content-length']) < CHUNK_SIZE):
```

Comment:
`int(resp.headers.get('content-length', CHUNK_SIZE + 1))` to prevent twice lookup.